db/compaction_picker.cc

```diff
@@ -1083,6 +1083,7 @@ void LevelCompactionBuilder::PickFilesMarkedForCompaction() {
 void LevelCompactionBuilder::SetupInitialFiles() {
   // Find the compactions by size on all levels.
   bool skipped_l0_to_base = false;
+  vstorage_->ComputeCompactionScore(ioptions_, mutable_cf_options_);
```

Comment:
I didn't quite get why this works. My understanding is compaction picking works like this:

1. `LevelCompactionPicker::NeedsCompaction` is called and checks if any level has score >= 1. If yes, we push the column family to the compaction queue
2. A background thread dequeues a column family from the compaction queue and calls `PickCompaction` on it. That function chooses which level to compact based on highest score.

I thought recomputing score (if we weren't doing it already) would have to be done before 1. as otherwise the levels might all have scores falsely below 1 (since score was last computed based on the assumption that the now-canceled compaction would successfully complete). But this is doing it in the beginning of 2. so I'm not sure how it helps compaction get rescheduled. Maybe we should debug :).

Reply:
So if the compaction gets cancelled, the column family is automatically pushed back into the compaction queue in db_impl_compaction_flush.cc, so PickCompaction is called at some point AFTER a compaction has been cancelled. However, PickCompaction was coming up with no compaction, because the compaction score of level0 was 0, because the PREVIOUS compaction picker *thought* it successfully compacted the 2 level0 files we have. Unfortunately, this compaction was cancelled, so this is a stale score.

Anyway, so if don't recompute compaction score here, the restarted, previously cancelled compaction, mistakenly has a stale score of 0.

Reply:
Oh ok, 1. is skipped because we use `AddToCompactionQueue` to requeue immediately. Let me think about it more. I suspect we should move it to after the sleep since, if there are multiple background threads, it could be picked up again immediately.

Reply:
OK, or we could just increment unscheduled_compactions_ in db_impl_compaction_flush.cc to make sure background functions know that a compaction needs to be done. But then, we still need to recompute the compaction score somewhere (perhaps in NeedsCompaction right before it checks CompactScore). Then cfd can be organically requeued. I don't actually know how much this buys us though. What exactly is the goal here? We just want to delay  when a cancelled compaction is rescheduled?

Reply:
I didn't have a goal above since I didn't fully understand yet. But now, yes, that is the primary goal. A secondary goal would be to put the call to recompute compaction score in the narrowest place where it's necessary. I agree it's idempotent, but the code is more understandable if we only call it in situations where compaction score could've changed.

I think a good place to put it is in the error handling code in `BackgroundCallCompaction`, after the sleep. It might even be worthwhile to move the `ReleaseCompactionFiles` there too, so other compaction threads don't reattempt the same failing work before the delay finishes. But I'll defer that decision to you.

Reply:
Hrm, it's actually quite difficult to both call ComputeCompactionScore and ReleaseCompactionFiles in the error handling code in BackgroundCallCompaction, because by then the compaction has been reset, so the cfd has been released. We can obviously pass up the cfd, if we encounter CompactionTooLarge status (I modified the status because I realized I created a status and never used it), but that seems inelegant.

So, I put ComputeCompactionScore right next to ReleaseCompactionFiles, because that's where you can access the cfd... If you have better ideas, I am open to them.

Reply:
OK that's fine for me. Another solution would be to move the sleep logic into `BackgroundCompaction` and refactor that function to have a single exit point. Then we can access cfd and release/recompute after the sleep. It's kind of tricky so one of us can try it in a follow-up PR if you agree it's reasonable.